br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -573,6 +590,115 @@ func (m *DupeDetector) RecordIndexConflictError(ctx context.Context, stream DupK
 	return nil
 }
 
+// RetrieveKeyAndValueFromErrFoundDuplicateKeys retrieves the key and value
+// from ErrFoundDuplicateKeys error.
+func RetrieveKeyAndValueFromErrFoundDuplicateKeys(err error) ([]byte, []byte, error) {
+	if !common.ErrFoundDuplicateKeys.Equal(err) {
+		return nil, nil, err
+	}
+	tErr, ok := errors.Cause(err).(*terror.Error)
+	if !ok {
+		return nil, nil, err
+	}
+	if len(tErr.Args()) != 2 {
+		return nil, nil, err
+	}
+	key, keyIsByte := tErr.Args()[0].([]byte)
+	value, valIsByte := tErr.Args()[1].([]byte)
+	if !keyIsByte || !valIsByte {
+		return nil, nil, err
+	}
+	return key, value, nil
+}
+
+// NewErrFoundConflictRecords generate an error ErrFoundDuplicateKeys
+// according to key and value.
+func NewErrFoundConflictRecords(ctx context.Context, key []byte, value []byte, tbl table.Table, idxInfo *model.IndexInfo) error {
+	logger := log.FromContext(ctx).With(zap.String("table", tbl.Meta().Name.L)).Begin(zap.InfoLevel, "[new-ErrFoundConflictRecords] new ErrFoundConflictRecords")
+	sessionOpts := encode.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+
+	logger2 := logger.With(zap.String("tableName", tbl.Meta().Name.L))
+	decoder, err := kv.NewTableKVDecoder(tbl, tbl.Meta().Name.L, &sessionOpts, logger2)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if tablecodec.IsRecordKey(key) {
+		// for data KV
+		handle, err := tablecodec.DecodeRowKey(key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		rowData := decoder.DecodeRawRowDataAsStr(handle, value)
+
+		return errors.Trace(common.ErrFoundDataConflictRecords.FastGenByArgs(tbl.Meta().Name, handle.String(), rowData))
+	}
+
+	// for index KV
+	_, idxID, _, err := tablecodec.DecodeIndexKey(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if idxInfo == nil {
+		idxInfo = model.FindIndexInfoByID(tbl.Meta().Indices, idxID)
+	}
+
+	h, err := decoder.DecodeHandleFromIndex(idxInfo, key, value)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rowKey := tablecodec.EncodeRowKeyWithHandle(tbl.Meta().ID, h)
+	handle, err := tablecodec.DecodeRowKey(rowKey)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	idxColLen := len(idxInfo.Columns)
+	indexName := fmt.Sprintf("%s.%s", tbl.Meta().Name.String(), idxInfo.Name.String())
+	colInfos := tables.BuildRowcodecColInfoForIndexColumns(idxInfo, tbl.Meta())
+	values, err := tablecodec.DecodeIndexKV(key, value, idxColLen, tablecodec.HandleNotNeeded, colInfos)
+	if err != nil {
+		logger.Warn("decode index key value failed", zap.String("index", indexName),
+			zap.String("key", hex.EncodeToString(key)), zap.String("value", hex.EncodeToString(value)), zap.Error(err))
+		return errors.Trace(common.ErrFoundIndexConflictRecords.FastGenByArgs(tbl.Meta().Name, key, value))
+	}
+	valueStr := make([]string, 0, idxColLen)
+	for i, val := range values[:idxColLen] {
+		d, err := tablecodec.DecodeColumnValue(val, colInfos[i].Ft, time.Local)
```

Comment:
I see these codes are the same as DDL's `genKeyExistsErr`, maybe ask @tangenta extract a function to avoid duplication. Also when DDL convert ErrFoundIndexConflictRecords to ErrKeyExists it can directly use `valueStr`